src/Components/Components/src/Routing/Router.cs

```diff
@@ -78,9 +88,11 @@ private string StringUntilAny(string str, char[] chars)
         /// <inheritdoc />
         protected virtual void Render(RenderTreeBuilder builder, Type handler, IDictionary<string, object> parameters)
         {
-            builder.OpenComponent(0, typeof(LayoutDisplay));
-            builder.AddAttribute(1, LayoutDisplay.NameOfPage, handler);
-            builder.AddAttribute(2, LayoutDisplay.NameOfPageParameters, parameters);
+            builder.OpenComponent(0, typeof(PageDisplay));
+            builder.AddAttribute(1, nameof(PageDisplay.Page), handler);
+            builder.AddAttribute(2, nameof(PageDisplay.PageParameters), parameters);
+            builder.AddAttribute(3, nameof(PageDisplay.NotAuthorizedContent), NotAuthorizedContent);
+            builder.AddAttribute(4, nameof(PageDisplay.AuthorizingContent), AuthorizingContent);
```

Comment:
This starts to feel a lot like "middleware". I think that longer term (before 3.0) we should find a better design that frees the router from these concerns.

Reply:
I felt exactly the same, and posted a proposal about this yesterday: https://github.com/aspnet/AspNetCore/issues/10493